### DISCUSSION TITLE
Is there a way to instantiate a Lightning Module class using a class method on Lightning CLI?

### DISCUSSION THREAD

**Discussion by carlesoctav** [Lightning-AI/pytorch-lightning#19863] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
Is there a way to instantiate a Lightning Module class using a class method on Lightning CLI? This problem arises when I try to create a model class that inherits from LightningModule and Huggingface's PretrainedModel Class. We all know that the PretrainedModel class uses the from_pretrained class method to instantiate a model loaded from the Huggingface hub. When I try to use LightningCLI, we all know that it will attempt to instantiate the instance with the __init__ method.

**Comment by mauvilsa** [comment]
With jsonargparse v4.30.0 just released, now it is possible to give in `class_path` the import path of a method, like `class_path: some.module.ClassName.method_name` (see [jsonargparse#513](https://github.com/omni-us/jsonargparse/pull/513)). But note, it is mandatory that the method have as return annotation the class type it returns. From what I have seen, this is not the case for Huggingface classes.

It will not be allowed to have functions without the return type annotation, because this would be vulnerability, since it would become possible to execute any function. But note, it shouldn't be too difficult to contribute required changes to [transformers](https://github.com/huggingface/transformers).

**Reply by carlesoctav** [reply]
Hi, I tried to add this feature but encountered the following error:

```
error: Parser key "model.model":
  SignatureArguments.add_function_arguments() got an unexpected keyword argument 'instantiate'
```

I added an annotation to the `PreTrainedModel` class because the `from_pretrained` method is located there (which every model inherits). Here's the code I used:

```python
class MLMTrainer(LightningModule):
    def __init__(
        self,
        model: PreTrainedModel,
        optimizer: OptimizerCallable,
        optimizer_config: CommonOptimizerConfig,
        lr_scheduler: LRSchedulerCallable,
        lr_scheduler_config: CommonLRSchedulerConfig,
        loss_fn: Optional[MLMLikeLossModule | Callable] = None,
        tokenizer_model_input: Optional[List[str]] = None,
        features_name: Optional[List[str]] = None,
        labels_name: Optional[List[str]] = None,
        **kwargs,
    ):
        super().__init__(**kwargs)
        self.optimizer = optimizer
        self.optimizer_config = optimizer_config
        self.model = model
        self.loss_fn = DefaultMLMLoss() if loss_fn is None else loss_fn
        self.lr_scheduler = lr_scheduler
        self.lr_scheduler_config = lr_scheduler_config
        self.tokenizer_model_input = tokenizer_model_input
        self.features_name = features_name
        self.labels_name = labels_name

    def training_step(self, *args: Any, **kwargs: Any):
        batch, batch_idx = args
        loss = self.step(batch, batch_idx, mode="train")
        self.log("train/loss", loss, on_step=True, on_epoch=True)
        return loss

    def step(self, batch, batch_idx, mode):
        features, labels = batch
        output = {}
        for key in self.features_name:
            output[key] = self.model(
                **{
                    f"{key}_{id}": features[f"{key}_{id}"]
                    for id in self.tokenizer_model_input
                }
            )
        loss = self.loss_fn(output, labels)

        self.log(f"{mode}/{self.loss_fn.__class__.__name__}", loss, on_step=True, on_epoch=True)
        return loss

    def validation_step(self, *args: Any, **kwargs: Any):
        batch, batch_idx = args
        loss = self.step(batch, batch_idx, mode="val")
        self.log("val/loss", loss, on_step=True, on_epoch=True)
        return loss

    def configure_optimizers(self):
        optimizer_config = self.optimizer_config.model_dump()
        lr_scheduler_config = self.lr_scheduler_config.model_dump()

        weight_decay = optimizer_config.pop("weight_decay")
        interval = lr_scheduler_config.pop("interval")
        frequency = lr_scheduler_config.pop("frequency")

        params = param_groups_weight_decay_default(self, weight_decay)
        optimizer = self.optimizer(params, **optimizer_config)

        scheduler = self.lr_scheduler(optimizer)
        scheduler_dict = {
            "scheduler": scheduler,
            "interval": interval,
            "frequency": frequency,
        }

        return {"optimizer": optimizer, "lr_scheduler": scheduler_dict}
        
        
def main():
    cli = LightningCLI(MLMTrainer, BoringDataModule)


if __name__ == "__main__":
    main()
```

And the configuration file:

```yaml
# lightning.pytorch==2.3.2
seed_everything: true
trainer:
  accelerator: auto
  strategy: auto
  devices: auto
  num_nodes: 1
  precision: null
  logger: null
  callbacks: null
  fast_dev_run: false
  max_epochs: null
  min_epochs: null
  max_steps: -1
  min_steps: null
  max_time: null
  limit_train_batches: null
  limit_val_batches: null
  limit_test_batches: null
  limit_predict_batches: null
  overfit_batches: 0.0
  val_check_interval: null
  check_val_every_n_epoch: 1
  num_sanity_val_steps: null
  log_every_n_steps: null
  enable_checkpointing: null
  enable_progress_bar: null
  enable_model_summary: null
  accumulate_grad_batches: 1
  gradient_clip_val: null
  gradient_clip_algorithm: null
  deterministic: null
  benchmark: null
  inference_mode: true
  use_distributed_sampler: true
  profiler: null
  detect_anomaly: false
  barebones: false
  plugins: null
  sync_batchnorm: false
  reload_dataloaders_every_n_epochs: 0
  default_root_dir: null
model:
  model:
    class_path: transformers.BertModel.from_pretrained
    init_args:
      pretrained_model_name_or_path: "bert-base-uncased"
  optimizer: torch.optim.Adam
  optimizer_config:
    lr: 2.0e-05
    weight_decay: 0
  lr_scheduler: torch.optim.lr_scheduler.ConstantLR
  lr_scheduler_config:
    interval: step
    frequency: 1
  loss_fn: null
  tokenizer_model_input: null
  features_name: null
  labels_name: null
ckpt_path: null
```

**Reply by carlesoctav** [reply]
also, i use jsonargparse==4.31.0

**Reply by mauvilsa** [reply]
I can take a look at it, thought not right now. What I can say now is that I see that the return type is `PreTrainedConfig` and the type hint of `model` is `PreTrainedModel`. Is `PreTrainedConfig` a subclass of `PreTrainedModel`? If yes, then this would be a bug in jsonargparse. If not, then the types seem wrong.

**Reply by carlesoctav** [reply]
i think `PreTrainedConfig` isn't a subclass of `PreTrainedModel`, but to initialized a PreTrainedModel you need to pass config, which type of config is `PreTraiendConfig. something like this:

```python
config = BertConfig()
model = BertModel(config)
```

**Reply by mauvilsa** [reply]
Created this https://github.com/huggingface/transformers/issues/31850

**Reply by mauvilsa** [reply]
Also created this pull request https://github.com/huggingface/transformers/pull/31869. However, I tested the bert model and I am still getting the same error. Seems that this case is somehow different to the feature that I implemented. Not sure how. Will try to look at this tomorrow.

**Reply by carlesoctav** [reply]
Do you think the problem lies in how Lightning implemented the CLI? I've never tried using "raw" jsonargparse cli before to build my own CLI.

**Reply by mauvilsa** [reply]
I haven't had much time to look at this. I did notice that there is some logic in jsonargparse that normalizes import paths, and it is not working correctly for `transformers.BertModel.from_pretrained`. Could be that this is the only problem or there are more.

**Reply by mauvilsa** [reply]
I did find another issue with the parameter resolver. So, not yet ready to work with transformer models.
